src/formatron/integrations/_utils.py

```diff
@@ -11,34 +11,34 @@ def _multiple_replace(replacements: typing.Dict[bytes, bytes], regex: re.Pattern
 Processors = set[typing.Literal["sentencepiece", "<0xHH>", "dot_G"]]
 
 
-def _autodetect_processors(vocab: typing.Dict[str, int]):
-    result = set()
+def _autodetect_processors(vocab: typing.Dict[str, int]) -> list[tuple[str, typing.Callable]]:
+    result = []
     llama_present = any(i.find('<0xF0>') != -1 for i in vocab.keys())
     underscore_present = (len([1 for i in vocab.keys() if i.find('\u2581') != -1]) / len(vocab)) > 0.2
     g_present = (len([1 for i in vocab.keys() if i.find('\u0120') != -1]) / len(vocab)) > 0.2
     if llama_present:
-        result.add("<0xHH>")
+        def update_vocab(old_char_to_new_char):
+            for j in range(256):
+                old_char_to_new_char[("<0x" + f"{j:02x}".upper() + ">").encode("UTF-8")] = bytes([j])
+        result.append(("<0xHH>", update_vocab))
     if underscore_present:
-        result.add("sentencepiece")
+        def update_vocab(old_char_to_new_char):
+            old_char_to_new_char["\u2581".encode("UTF-8")] = b" "
+        result.append(("sentencepiece", update_vocab))
     elif g_present:
-        result.add("dot_G")
+        def update_vocab(old_char_to_new_char):
+            old_char_to_new_char.update(huggingface_bytelevel_decoder())
+        result.append(("dot_G", update_vocab))
     return result
 
 
-def get_original_characters(vocab: typing.Dict[str, int]) -> typing.Dict[int, bytes]:
+def get_original_characters(vocab: typing.Dict[str, int], processors: typing.Optional[list[tuple[str, typing.Callable]]] = None) -> typing.Dict[int, bytes]:
     old_char_to_new_char = {}
     assert len(set(vocab.values())) == len(vocab), "Vocabulary contains duplicate token IDs!"
-    processors = _autodetect_processors(vocab)
-    for i in processors:
-        if i == "sentencepiece":
-            old_char_to_new_char["\u2581".encode("UTF-8")] = b" "
-        elif i == "dot_G":
-            old_char_to_new_char.update(huggingface_bytelevel_decoder())
-        elif i == "<0xHH>":
-            for j in range(256):
-                old_char_to_new_char[("<0x" + f"{j:02x}".upper() + ">").encode("UTF-8")] = bytes([j])
-        else:
-            raise ValueError(f"{i} is not a valid processor name!")
+    if processors is None:
```

Comment:
One thing I am considering is that people probably still want some kinds of default autodetection.  Maybe we should make _autodetect_processors public so people can use it when necessary? Or we should extract the "sentencepiece", "dot_G", "<0xHH>" into separate public functions from _autodetect_processors? Or maybe both? All of them sound reasonable from my standpoint, but I am unsure how the thing will look like from a user's perspective.

Reply:
Sounds reasonable to me, but also unsure how people would use that. In my case, I had preferred to explicitly provide those callables. Having said that, I don't see harm in extracting those update function, which I now did in the latest commit.

Reply:
I agree. They look good and I think we just need to add some doc strings to these new methods to make users aware about their usage.